server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/allocation/decider/WriteLoadConstraintDeciderIT.java

```diff
@@ -264,13 +269,53 @@ public void testShardsAreAssignedToNotPreferredWhenAlternativeIsNo() {
         }
     }
 
+    @TestLogging(
+        reason = "track when reconciliation has completed",
+        value = "org.elasticsearch.cluster.routing.allocation.allocator.DesiredBalanceShardsAllocator:DEBUG"
+    )
+    public void testAllocationExplainMoveShardNotPreferred() {
+        TestHarness harness = setUpThreeTestNodesAndAllIndexShardsOnFirstNode();
+
+        // Running the {@link #runCanRemainNotPreferredIsIgnoredWhenAllOtherNodesReturnNotPreferred} logic should set up a cluster where
+        // shards are all allocated to a {@link AllocationDeciders#canRemain} {@link Decision#NOT_PREFERRED} node, while the other nodes
+        // return {@link AllocationDeciders#canAllocate} {@link Decision#NOT_PREFERRED} responses. This should exercise NOT_PREFERRED in
+        // the allocation/explain paths for remaining on a node AND assignment to other nodes.
+        runCanRemainNotPreferredIsIgnoredWhenAllOtherNodesReturnNotPreferred(harness);
+
+        ClusterAllocationExplainRequest allocationExplainRequest = new ClusterAllocationExplainRequest(
+            TEST_REQUEST_TIMEOUT,
+            harness.indexName,
+            0,
+            true,
+            null
+        );
+        var allocationExplainResponse = safeGet(client().execute(TransportClusterAllocationExplainAction.TYPE, allocationExplainRequest));
+        logger.info("---> Allocation explain response: " + Strings.toString(allocationExplainResponse.getExplanation(), true, true));
+
+        var decision = allocationExplainResponse.getExplanation().getShardAllocationDecision().getMoveDecision();
+        assertThat("Rebalancing should be disabled", decision.canRebalanceCluster(), equalTo(false));
+        assertThat(decision.getCanRemainDecision().type(), equalTo(Decision.NOT_PREFERRED.type()));
+        assertNull(decision.getTargetNode());
+        assertThat(decision.getAllocationDecision(), equalTo(AllocationDecision.NOT_PREFERRED));
+
+        var canAllocateDecisions = allocationExplainResponse.getExplanation()
+            .getShardAllocationDecision()
+            .getMoveDecision()
+            .getNodeDecisions();
+        assertThat(canAllocateDecisions.size(), equalTo(2));
```

Comment:
can we give `2` a name somewhere? I assume it's `{number of cluster nodes} - 1` perhaps we should make our `expectedNumberOfClusterNodes` explicit and assert that it's accurate after we run the other test logic?

Reply:
Yep, updated to indicate purpose and make dynamic 👍